File: MAXED/prt.pl
#!/usr/bin/perl

sub PRT1{
#  This subroutine prints intermediate output, as does PRT2 through
#  PRT10. Note that if SA is minimizing the function, the sign of the
#  function value and the directions (up/down) are reversed in all
#  output to correspond with the actual function optimization. This
#  correction is because SA was written to maximize functions and
#  it minimizes by maximizing the negative a function.

	print "  THE STARTING VALUE (X) IS OUTSIDE THE BOUNDS ",
		br,"  (LB AND UB). EXECUTION TERMINATED WITHOUT ANY",
		br,"  OPTIMIZATION. RESPECIFY X, UB OR LB SO THAT  ",
		br,"  LB(I) .LT. X(I) .LT. UB(I), I = 1, N. ",br;

	return;
}

sub PRT2{
	local *MAX=shift;
	local *N=shift;
	local *X=shift;
	local *F=shift;

	print "   ";
    &PRTVEC(\@X,$N,'INITIAL X');
	print "\n";
	if($MAX) {
		printf "  INITIAL F: %25.18g",$F;
	}
	else{
		printf "  INITIAL F: %25.18g",-$F;
	}

	return;
}

sub PRT3{
	local *MAX=shift;
	local *N=shift;
	local *XP=shift;
	local *X=shift;
	local *FP=shift;
	local *F=shift;

	print "   ";
    &PRTVEC(\@X,$N,'CURRENT X');
	if($MAX) {
		printf "  CURRENT F: %25.18g",$F;
	}
	else{
		printf "  CURRENT F: %25.18g",-$F;
	}
    &PRTVEC(\@XP,$N,'TRIAL X');
    print "  POINT REJECTED SINCE OUT OF BOUNDS";
	return;
}

      

sub PRT4{
	local *MAX=shift;
	local *N=shift;
	local *XP=shift;
	local *X=shift;
	local *FP=shift;
	local *F=shift;
	
    print "   ";
	&PRTVEC(\@X,$N,'CURRENT X');
	if($MAX) {
		printf "  CURRENT F: %25.18g",$F;
		&PRTVEC(\$XP,$N,'TRIAL X');
		printf "  RESULTING F: %25.18g",$FP;
	}
	else{
		printf "  CURRENT F: %25.18g",-$F;
		&PRTVEC(\$XP,\$N,'TRIAL X');
		printf "  RESULTING F: %25.18g",-$FP;
	}
	return;
}


      
sub PRT5{

	print  "  TOO MANY FUNCTION EVALUATIONS; CONSIDER ",
		br,"  INCREASING MAXEVL OR EPS, OR DECREASING ",
		br,"  NT OR RT. THESE RESULTS ARE LIKELY TO BE ",
		br,"  POOR.";
	return;
}


sub PRT6{
	local *MAX=shift;

	if ($MAX){
		print "  THOUGH LOWER, POINT ACCEPTED";
	}
	else {
		print "  THOUGH HIGHER, POINT ACCEPTED";
	}

	return;
}

sub PRT7{
	local *MAX=shift;

	if ($MAX){
		print "  LOWER POINT REJECTED";
	}
	else {
		print "  HIGHER POINT REJECTED";
	}
	return;
}


sub PRT8{
	local *N=shift;
	local *VM=shift;
	local *XOPT=shift;
	local *X=shift;

	print "\n";
	print "INTERMEDIATE RESULTS AFTER STEP LENGTH ADJUSTMENT";
	print "\n";
#	print br;
	&PRTVEC(\@VM,$N,'NEW STEP LENGTH (VM)');
    &PRTVEC(\@XOPT,$N,'CURRENT OPTIMAL X');
    &PRTVEC(\@X,$N,'CURRENT X');
    print br;
	return;
}

sub PRT9{
	local *MAX=shift;
	local *N=shift;
	local *T=shift;
	local *XOPT=shift;
	local *VM=shift;
	local *FOPT=shift;
	local *NUP=shift;
	local *NDOWN=shift;
	local *NREJ=shift;
	local *LNOBDS=shift;
	local *NNEW=shift;

	my $TOTMOV;

	$TOTMOV = $NUP + $NDOWN + $NREJ;

	print "\n\n";
	print  "  INTERMEDIATE RESULTS BEFORE NEXT TEMPERATURE REDUCTION";
	print "\n\n";
	print br;
	printf "  CURRENT TEMPERATURE:            %12.5g",$T;
	print "\n";
	print br;

	if ($MAX){
		printf "  MAX FUNCTION VALUE SO FAR:  %25.18g",$FOPT;
		print "\n";
#		print br;
		printf "  TOTAL MOVES:                %8d",$TOTMOV;
		print "\n";
#		print br;
	 	printf "  UPHILL:                     %8d",$NUP;
		print "\n";
#		print br;
		printf "     ACCEPTED DOWNHILL:       %8d", $NDOWN;
		print "\n";
#		print br;
		printf "     REJECTED DOWNHILL:       %8d", $NREJ;
		print "\n";
#		print br;
		printf "  OUT OF BOUNDS TRIALS:       %8d", $LNOBDS;
		print "\n";
#		print br;
		printf "  NEW MAXIMA THIS TEMPERATURE:%8d", $NNEW;
		print "\n";
#		print br;
	}
	else {
		printf "  MIN FUNCTION VALUE SO FAR:  %25.18g", -$FOPT;
#		print br;
		print "\n";
		printf "  TOTAL MOVES:                %8d", $TOTMOV;
		print "\n";
#		print br;
		printf "     DOWNHILL:                %8d", $NUP;
		print "\n";
#		print br;
		printf "     ACCEPTED UPHILL:         %8d", $NDOWN;
		print "\n";
#		print br;
		printf "     REJECTED UPHILL:         %8d", $NREJ;
		print "\n";
#		print br;
		printf "  TRIALS OUT OF BOUNDS:       %8d", $LNOBDS;
		print "\n";
#		print br;
		printf "  NEW MINIMA THIS TEMPERATURE:%8d", $NNEW;
		print "\n";
#		print br;
	}
	&PRTVEC(\@XOPT,$N,'CURRENT OPTIMAL X');
    &PRTVEC(\@VM,$N,'STEP LENGTH (VM)');
	print br;
	return;
}

sub PRT10{

	printf "  SA ACHIEVED TERMINATION CRITERIA. IER = 0. ",br;
	return;
}
1;

File: bums2/FORTRAN_METHODS/MAXED/annealing/prt.py
#This is the python version of prt.pl
#prt.pl operates as an SA intermediate output reported

#  This subroutine prints intermediate output, as does PRT2 through
#  PRT10. Note that if SA is minimizing the function, the sign of the
#  function value and the directions (up/down) are reversed in all
#  output to correspond with the actual function optimization. This
#  correction is because SA was written to maximize functions and
#  it minimizes by maximizing the negative a function.

from .prtvec import VectorPrinter

class SAReporter:
    #Prints the simulated annealing intermediate messages, all the methods should correspond if they share a name
    @staticmethod
    def prt1(i, xi, LB, UB):
        print(f"  THE STARTING VALUE {xi} IS OUTSIDE THE BOUNDS")
        print("  (LB AND UB). EXECUTION TERMINATED WITHOUT ANY")
        print("  OPTIMIZATION. RESPECIFY X, UB OR LB SO THAT")
        print(f"  {LB} < {xi} < {UB}, I = {i}, N.")

    @staticmethod
    def prt2(maximize, x, f):
        print()
        VectorPrinter.print_vector("INITIAL X", x)
        print(f"  INITIAL F: {f if maximize else -f:.18g}")

    @staticmethod
    def prt3(maximize, xp, x, f):
        print()
        VectorPrinter.print_vector("CURRENT X", x)
        print(f"  CURRENT F: {f if maximize else -f:.18g}")
        VectorPrinter.print_vector("TRIAL X", xp)
        print("  POINT REJECTED SINCE OUT OF BOUNDS")

    @staticmethod
    def prt4(maximize, xp, x, fp, f):
        VectorPrinter.print_vector("CURRENT X", x)
        print(f"  CURRENT F: {f if maximize else -f:.18g}")
        VectorPrinter.print_vector("TRIAL X", xp)
        print(f"  RESULTING F: {fp if maximize else -fp:.18g}")

    @staticmethod
    def prt5():
        print("  TOO MANY FUNCTION EVALUATIONS; CONSIDER")
        print("  INCREASING MAXEVL OR EPS, OR DECREASING")
        print("  NT OR RT. THESE RESULTS ARE LIKELY TO BE")
        print("  POOR.")

    @staticmethod
    def prt6(maximize):
        if maximize:
            print("  THOUGH LOWER, POINT ACCEPTED")
        else:
            print("  THOUGH HIGHER, POINT ACCEPTED")

    @staticmethod
    def prt7(maximize):
        if maximize:
            print("  LOWER POINT REJECTED")
        else:
            print("  HIGHER POINT REJECTED")

    @staticmethod
    def prt8(vm, xopt, x):
        print("\nINTERMEDIATE RESULTS AFTER STEP LENGTH ADJUSTMENT\n")
        VectorPrinter.print_vector("NEW STEP LENGTH (VM)", vm)
        VectorPrinter.print_vector("CURRENT OPTIMAL X", xopt)
        VectorPrinter.print_vector("CURRENT X", x)
        print()

    @staticmethod
    def prt9(
        maximize, n, t, xopt, vm, fopt,
        nup, ndown, nrej, out_of_bounds, nnew
    ):
        totmov = nup + ndown + nrej
        print("\n\nINTERMEDIATE RESULTS BEFORE NEXT TEMPERATURE REDUCTION")
        print(f"  CURRENT TEMPERATURE: {t}")
        if maximize:
            print(f"  MAX FUNCTION VALUE SO FAR: {fopt:.18g}")
            print(f"  TOTAL MOVES: {totmov}")
            print(f"  UPHILL: {nup}")
            print(f"     ACCEPTED DOWNHILL: {ndown}")
            print(f"     REJECTED DOWNHILL: {nrej}")
            print(f"  OUT OF BOUNDS TRIALS: {out_of_bounds}")
            print(f"  NEW MAXIMA THIS TEMPERATURE: {nnew}")
        else:
            print(f"  MIN FUNCTION VALUE SO FAR: {-fopt:.18g}")
            print(f"  TOTAL MOVES: {totmov}")
            print(f"     DOWNHILL: {nup}")
            print(f"     ACCEPTED UPHILL: {ndown}")
            print(f"     REJECTED UPHILL: {nrej}")
            print(f"  TRIALS OUT OF BOUNDS: {out_of_bounds}")
            print(f"  NEW MINIMA THIS TEMPERATURE: {nnew}")
        VectorPrinter.print_vector("CURRENT OPTIMAL X", xopt)
        VectorPrinter.print_vector("STEP LENGTH (VM)", vm)

    @staticmethod
    def prt10():
        print("  SA ACHIEVED TERMINATION CRITERIA. IER = 0.")